Add ability to exclude more resources via the cloud-nuke-excuded tag

merkata opened the pull request:
Currently, you can only exclude S3 buckets that are tagged with key cloud-nuke-excluded and value set to true. This PR adds more resources that are protected via this tag. This functionality is used in case you have a test environment that you do not want to get wiped out, like an integration setup with an EKS cluster, ELB, workers in an ASG etc. - you can propagate the tag across your setup for more fine-grained control (in comparison to specifying a creation time). The current setup is backwards compatible and has been tested with every resource.

harmrai-pol commented:
Nice work @merkata! What is preventing this from getting merged in @brikis98 @yorinasub17

gmembre-zenika commented:
It would  be great if this feature can be reviewed / merged 💚

tommy31 commented:
Bump.

This PR is related to this issue [166](https://github.com/gruntwork-io/cloud-nuke/issues/166)

gforien commented:
Hi 👋 
Any updates on this ?

ellisonc commented:
@merkata this change set looks great to me, could you resolve the conflicts in the PR and add test cases for the new tagging feature?

merkata commented:
> @merkata this change set looks great to me, could you resolve the conflicts in the PR and add test cases for the new tagging feature?

Thanks a lot @ellisonc  - I've fixed the conflicts and I'll be writing the tests to cover the new functionality. Appreciate your input!

zackproser commented:
> > @merkata this change set looks great to me, could you resolve the conflicts in the PR and add test cases for the new tagging feature?
> 
> Thanks a lot @ellisonc - I've fixed the conflicts and I'll be writing the tests to cover the new functionality. Appreciate your input!

@merkata This looks like a great change! Could you also please update the README to reflect this expanded functionality as well?

merkata commented:
> > > @merkata this change set looks great to me, could you resolve the conflicts in the PR and add test cases for the new tagging feature?
> > 
> > 
> > Thanks a lot @ellisonc - I've fixed the conflicts and I'll be writing the tests to cover the new functionality. Appreciate your input!
> 
> @merkata This looks like a great change! Could you also please update the README to reflect this expanded functionality as well?

Thanks @zackproser , I'll update the README as well!

ellisonc commented:
@merkata apologies for the delay, I attempted to fix the conflicts and get this merged but our tooling doesn't allow pushes to any branch named master and I don't want to steal your credit by making the changes on a new branch of my own.  Could you resolve the conflicts on a feature branch and update the PR?

merkata commented:
> @merkata apologies for the delay, I attempted to fix the conflicts and get this merged but our tooling doesn't allow pushes to any branch named master and I don't want to steal your credit by making the changes on a new branch of my own. Could you resolve the conflicts on a feature branch and update the PR?

Hey @ellisonc , thanks for the feedback! I fixed the conflicts, two things stand out:

* the current `README.md` doesn't seem to cover exclusion of resources based on tags, the previous list of resources described which resources can be filtered out based on tag and I extended the descriptions of the resources I added.

* please have a look at `aws/snapshot.go`, I believe the logic is correct, still there is the addition of not deleting a snapshot with the AWS backup tag and I want to make sure I got the logic right.

Thanks!

ellisonc commented on aws/nat_gateway.go:
```diff
@@ -54,7 +54,11 @@ func shouldIncludeNatGateway(ngw *ec2.NatGateway, excludeAfter time.Time, config
+	if hasNGWExcludeTag(ngw) {
+=======
 	if aws.StringValue(ngw.State) == ec2.NatGatewayStateDeleted {
+>>>>>>> upstream/master
```
Weirdness from merge conflict here

merkata replied:
Totally missed that one, sorry! Fixed.

ellisonc reviewed (commented):
@merkata the logic all looks good, just some leftover issues from the merge conflict fix

ellisonc commented:
@merkata I also don't see the readme update you mentioned, did that not get staged?

merkata commented:
> @merkata I also don't see the readme update you mentioned, did that not get staged?

I didn't explain it accurately - I updated the old version of the README.md, it looked like this:

```text
Inspecting and deleting all ACM Private CA in an AWS account
Inspecting and deleting all Auto scaling groups in an AWS account - except for ASGs tagged with Key=cloud-nuke-excluded Value=true
Inspecting and deleting all Elastic Load Balancers (v1 and v2) in an AWS account - except for ELBs tagged with Key=cloud-nuke-excluded Value=true
Inspecting and deleting all Transit Gateways in an AWS account
Inspecting and deleting all EBS Volumes in an AWS account - except for volumes tagged with Key=cloud-nuke-excluded Value=true
...
```

and I specifically added the _except for \<resource name\> tagged with..._ section for each resource.

Now the section looks like this:

```text
Cloud-nuke suppports 🔎 inspecting and 🔥💀 deleting the following AWS resources:

| Resource Family | Resource type 
| --------------- | ---------- 
| EC2 | Auto scaling groups |
| EC2 | Elastic Load Balancers (v1 and v2) |
| EC2 | EBS Volumes | 
| EC2 | Unprotected EC2 instances |
...
```

So I'm not sure where to make a note about excluding a resource based on an exclusion tag.

merkata commented:
> @merkata I see what you're saying. How about adding a new section header under the [`Excluding Resources by Age`](https://github.com/gruntwork-io/cloud-nuke#excluding-resources-by-age) header (or maybe right above it?)

@zackproser sorry I wasn't able to response earlier, I've added a short paragraph with the list of resources that would be supported, let me know what you think of it. Thanks!

zackproser commented on aws/asg.go:
```diff
@@ -32,6 +32,17 @@ func getAllAutoScalingGroups(session *session.Session, region string, excludeAft
 	return groupNames, nil
 }
 
+// hasASGExcludeTag checks whether the exlude tag is set for a resource to skip deleting it.
```
```suggestion
// hasASGExcludeTag checks whether the exclude tag is set for a resource to skip deleting it.
```

zackproser commented on aws/ebs.go:
```diff
@@ -43,6 +43,17 @@ func getAllEbsVolumes(session *session.Session, region string, excludeAfter time
 	return volumeIds, nil
 }
 
+// hasEBSExcludeTag checks whether the exlude tag is set for a resource to skip deleting it.
```
```suggestion
// hasEBSExcludeTag checks whether the exclude tag is set for a resource to skip deleting it.
```

zackproser commented on aws/ebs.go:
```diff
@@ -43,6 +43,17 @@ func getAllEbsVolumes(session *session.Session, region string, excludeAfter time
+func hasEBSExcludeTag(volume *ec2.Volume) bool {
+	// Exclude deletion of any buckets with cloud-nuke-excluded tags
+	for _, tag := range volume.Tags {
+		if *tag.Key == AwsResourceExclusionTagKey && *tag.Value == "true" {
```
NIT - In general we want to use the `aws.StringValue` and related methods for dereferencing pointers - partly for consistency, but mainly for the fact that these methods include a safety check to ensure we're not dereferencing a nil pointer.

zackproser commented on aws/eip.go:
```diff
@@ -82,6 +82,17 @@ func getAllEIPAddresses(session *session.Session, region string, excludeAfter ti
 	return allocationIds, nil
 }
 
+// hasEIPExcludeTag checks whether the exlude tag is set for a resource to skip deleting it.
```
```suggestion
// hasEIPExcludeTag checks whether the exclude tag is set for a resource to skip deleting it.
```

zackproser commented on aws/elbv2.go:
```diff
@@ -18,22 +18,41 @@ import (
 }
 
-func shouldIncludeELBv2(balancer *elbv2.LoadBalancer, excludeAfter time.Time, configObj config.Config) bool {
+// hasELBv2ExcludeTag checks whether the exlude tag is set for a resource to skip deleting it.
```
```suggestion
// hasELBv2ExcludeTag checks whether the exclude tag is set for a resource to skip deleting it.
```

zackproser commented on aws/iam.go:
```diff
@@ -32,14 +32,27 @@ func getAllIamUsers(session *session.Session, excludeAfter time.Time, configObj
 	return userNames, nil
 }
 
+// hasIAMExcludeTag checks whether the exlude tag is set for a resource to skip deleting it.
```
```suggestion
// hasIAMExcludeTag checks whether the exclude tag is set for a resource to skip deleting it.
```

zackproser commented on aws/nat_gateway.go:
```diff
@@ -65,6 +70,17 @@ func shouldIncludeNatGateway(ngw *ec2.NatGateway, excludeAfter time.Time, config
 	)
 }
 
+// hasNGWExcludeTag checks whether the exlude tag is set for a resource to skip deleting it.
```
```suggestion
// hasNGWExcludeTag checks whether the exclude tag is set for a resource to skip deleting it.
```

zackproser commented on aws/rds.go:
```diff
@@ -34,6 +37,17 @@ func getAllRdsInstances(session *session.Session, excludeAfter time.Time, config
 	return names, nil
 }
 
+// hasRDSExcludeTag checks whether the exlude tag is set for a resource to skip deleting it.
```
```suggestion
// hasRDSExcludeTag checks whether the exclude tag is set for a resource to skip deleting it.
```

zackproser commented on aws/rds_cluster.go:
```diff
@@ -45,14 +45,27 @@ func getAllRdsClusters(session *session.Session, excludeAfter time.Time) ([]*str
 	return names, nil
 }
 
+// hasRDSClusterExcludeTag checks whether the exlude tag is set for a resource to skip deleting it.
```
```suggestion
// hasRDSClusterExcludeTag checks whether the exclude tag is set for a resource to skip deleting it.
```

zackproser commented on aws/snapshot.go:
```diff
@@ -37,14 +37,27 @@ func getAllSnapshots(session *session.Session, region string, excludeAfter time.
 	return snapshotIds, nil
 }
 
+// hasEBSSnapExcludeTag checks whether the exlude tag is set for a resource to skip deleting it.
```
```suggestion
// hasEBSSnapExcludeTag checks whether the exclude tag is set for a resource to skip deleting it.
```

zackproser commented:
This is looking very close @merkata, thanks for the great work! Left a couple NITs (one typo that was present in a couple of places, and some thoughts on the use of `aws.StringValue` and related methods for dereferencing pointers. Otherwise we should be good to get this merged shortly :+1:

hongil0316 commented:
Hey @merkata, we haven't heard back from you for quite some time. Do you plan to continue making progress on this PR? 
We've recently made some refactoring changes in the code and it may require a bit of conflict resolution.

hongil0316 commented:
Moved this change to this PR - https://github.com/gruntwork-io/cloud-nuke/pull/573.
